aggregator_core/src/task.rs

```diff
@@ -481,6 +481,486 @@ impl Task {
+        }
+
+        Ok(())
+    }
+
+    /// Returns the [`VerifyKey`] used by this aggregator to prepare report shares with other
+    /// aggregators.
+    ///
+    /// # Errors
+    ///
+    /// If the verify key is not the correct length as required by the VDAF, an error will be
+    /// returned.
+    pub fn vdaf_verify_key<const SEED_SIZE: usize>(&self) -> Result<VerifyKey<SEED_SIZE>, Error> {
+        VerifyKey::try_from(&self.vdaf_verify_key).map_err(|_| Error::AggregatorVerifyKeySize)
+    }
+}
+
+/// An aggregator's view of the task's parameters.
+#[derive(Clone, Derivative, PartialEq, Eq)]
+#[derivative(Debug)]
+pub struct AggregatorTask {
+    /// Common task parameters
+    common_parameters: CommonTaskParameters,
+    /// URL relative to which the peer aggregator's API endpoints are found.
+    #[derivative(Debug(format_with = "std::fmt::Display::fmt"))]
+    peer_aggregator_endpoint: Url,
+    /// Parameters specific to either aggregator role
+    aggregator_parameters: AggregatorTaskParameters,
+    /// HPKE configurations & private keys used by this aggregator to decrypt client reports.
+    hpke_keys: HashMap<HpkeConfigId, HpkeKeypair>,
+}
+
+impl AggregatorTask {
+    /// Create a new [`AggregatorTask`] with the provided values.
+    #[allow(clippy::too_many_arguments)]
+    pub fn new<I: IntoIterator<Item = HpkeKeypair>>(
+        task_id: TaskId,
+        peer_aggregator_endpoint: Url,
+        query_type: QueryType,
+        vdaf: VdafInstance,
+        vdaf_verify_key: SecretBytes,
+        max_batch_query_count: u64,
+        task_expiration: Option<Time>,
+        report_expiry_age: Option<Duration>,
+        min_batch_size: u64,
+        time_precision: Duration,
+        tolerable_clock_skew: Duration,
+        hpke_keys: I,
+        aggregator_parameters: AggregatorTaskParameters,
+    ) -> Result<Self, Error> {
+        let common_parameters = CommonTaskParameters::new(
+            task_id,
+            query_type,
+            vdaf,
+            vdaf_verify_key,
+            max_batch_query_count,
+            task_expiration,
+            report_expiry_age,
+            min_batch_size,
+            time_precision,
+            tolerable_clock_skew,
+        );
+        Self::new_with_common_parameters(
+            common_parameters,
+            peer_aggregator_endpoint,
+            hpke_keys,
+            aggregator_parameters,
+        )
+    }
+
+    fn new_with_common_parameters<I: IntoIterator<Item = HpkeKeypair>>(
+        common_parameters: CommonTaskParameters,
+        peer_aggregator_endpoint: Url,
+        hpke_keys: I,
+        aggregator_parameters: AggregatorTaskParameters,
+    ) -> Result<Self, Error> {
+        // Compute hpke_configs mapping cfg.id -> (cfg, key).
+        let hpke_keys: HashMap<HpkeConfigId, HpkeKeypair> = hpke_keys
+            .into_iter()
+            .map(|keypair| (*keypair.config().id(), keypair))
+            .collect();
+
+        common_parameters.validate()?;
```

Comment:
It seems out of place that this is in a constructor instead of `AggregatorTask::validate()`, or instead of having the constructor check all validation constraints.

It looks like our old `Task` struct didn't validate on construction largely because it was internally used as storage inside `TaskBuilder`, so we intentionally initialized it with inconsistent values, wrote to fields without re-checking invariants, etc., and then only validated parameters on conversion from a `TaskBuilder` into the inner `Task`. Since `AggregatorTask` doesn't get used for that purpose anymore, we could make the `validate()` methods private, make both constructors fallible, and always validate before returning a `CommonTaskParameters` or `AggregatorTask`.

Reply:
I folded the validate methods into the constructors, since validation never needs to be done independently of instantiation of a `CommonTaskParameters` or `AggregatorTask` value.